packages/internal/bridge/sdk/src/tokenBridge.ts

```diff
@@ -547,143 +477,640 @@ export class TokenBridge {
-    token: string,
-    senderOrRecipientAddress: string,
-    amount: ethers.BigNumber,
-    sourceChainId: string,
+  private async getDynamicWithdrawGas(
     destinationChainId: string,
-  ) {
-    if (!ethers.utils.isAddress(senderOrRecipientAddress)) {
-      throw new BridgeError(
-        `address ${senderOrRecipientAddress} is not a valid address`,
-        BridgeErrorType.INVALID_ADDRESS,
+    sender: string,
+    recipient: string,
+    token: FungibleToken,
+    amount: ethers.BigNumber,
+  ): Promise<number> {
+    let rootToken: string;
+    if (token.toUpperCase() === NATIVE
+      || this.isWrappedIMX(token, destinationChainId)) {
+      rootToken = this.getRootIMX(destinationChainId);
+    } else {
+      // Find root token
+      const erc20Contract: ethers.Contract = await withBridgeError<ethers.Contract>(
+        async () => new ethers.Contract(token, CHILD_ERC20, this.config.childProvider),
+        BridgeErrorType.PROVIDER_ERROR,
       );
+      rootToken = await withBridgeError<Address>(() => erc20Contract.rootToken(), BridgeErrorType.PROVIDER_ERROR);
     }
+    // Encode payload
+    const payload = defaultAbiCoder.encode(
+      ['bytes32', 'address', 'address', 'address', 'uint256'],
+      [WITHDRAW_SIG, rootToken, sender, recipient, amount],
+    );
+    // Generate unique command ID based on payload and current time.
+    const commandId = keccak256(
+      defaultAbiCoder.encode(['bytes', 'uint256'], [payload, new Date().getTime()]),
+    );
+    const sourceChain = this.getChildchain(destinationChainId);
+    const sourceAddress = ethers.utils.getAddress(this.getChildAdaptor(destinationChainId)).toString();
+    const destinationAddress = this.getRootAdaptor(destinationChainId);
+    const payloadHash = keccak256(payload);
+    // Calculate slot key for given command ID.
+    const command = defaultAbiCoder.encode(
+      ['bytes32', 'bytes32', 'string', 'string', 'address', 'bytes32'],
+      [SLOT_PREFIX_CONTRACT_CALL_APPROVED, commandId, sourceChain, sourceAddress, destinationAddress, payloadHash],
+    );
+    const commandHash = keccak256(command);
+    const slot = keccak256(concat([commandHash, hexlify(zeroPad(hexlify(SLOT_POS_CONTRACT_CALL_APPROVED), 32))]));
 
-    // The deposit amount cannot be <= 0
-    if (amount.isNegative() || amount.isZero()) {
+    // Encode execute data
+    const axelarAdapterContract: ethers.Contract = await withBridgeError<ethers.Contract>(
+      async () => new ethers.Contract(destinationAddress, ROOT_AXELAR_ADAPTOR, this.config.rootProvider),
+      BridgeErrorType.PROVIDER_ERROR,
+    );
+    const executeData = await withBridgeError<string>(
+      async () => axelarAdapterContract.interface
+        .encodeFunctionData('execute', [commandId, sourceChain, sourceAddress, payload]),
+      BridgeErrorType.INTERNAL_ERROR,
+    );
+
+    // Build simulation
+    const axelarGateway = this.getAxelarGateway(destinationChainId);
+    const simulations = [{
```

Comment:
Not sure if any functions could be written to reduce the code duplication with here and the simulation stuff in getDynamicDepositGas